### PR TITLE
Support GCS

### LINKED ISSUE #3760: Support Google Cloud Storage in UI
### My actions before raising this issue
- [x] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [x] Searched [past issues](/issues)

### Current Behaviour

Only a [server](https://github.com/openvinotoolkit/cvat/pull/3561) support for `Google Cloud Storage` was implemented.
### Possible Solution

Need to extend current implementation for cloud storages and add support for `Google Cloud Storage` in UI

### PR DESCRIPTION
**Pull request by Marishka17**
### Motivation and context

Resolve #3760 

### How has this been tested?

Manually + extended jest tests
### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [x] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
~~- [ ] I have updated the license header for each file (see an example below)~~

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: cvat/apps/engine/models.py [hunk 1]
```diff
@@ -583,23 +583,27 @@ def __str__(self):
...
     # AWS temporary session tocken - None
     # The size of the security token that AWS STS API operations return is not fixed.
     # We strongly recommend that you make no assumptions about the maximum size.
     # The typical token size is less than 4096 bytes, but that can vary.
+    # specific attributes:
+    # location - max 23
+    # project ID: 6 - 30 (https://cloud.google.com/resource-manager/docs/creating-managing-projects#before_you_begin)
     provider_type = models.CharField(max_length=20, choices=CloudProviderChoice.choices())
-    resource = models.CharField(max_length=63)
+    resource = models.CharField(max_length=222)
```
Comment (Marishka17) [inline cvat/apps/engine/models.py:598]: @nmanovic, I've increased restrictions for bucket names length because google cloud storage has 2 restriction [variants](https://cloud.google.com/storage/docs/naming-buckets#requirements) - 63 like other providers and 222 with dots in bucket name.  Are we going to support the second case?

File: cvat-core/src/server-proxy.js [hunk 2]
```diff
@@ -1181,28 +1181,42 @@
...
+                const storageDetailData = new FormData();
+                for (const [key, value] of Object.entries(storageDetail)) {
+                    if (Array.isArray(value)) {
+                        value.forEach((element, idx) => {
+                            storageDetailData.append(`${key}[${idx}]`, element);
+                        });
+                    } else {
+                        storageDetailData.set(key, value);
+                    }
+                }
```
Comment (bsekachev) [inline cvat-core/src/server-proxy.js:1216]: The same code like in ``createCloudStorage``

File: cvat-ui/src/assets/google-cloud.svg [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<svg width="1em" height="1em" viewBox="0 0 256 206" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (bsekachev) [inline cvat-ui/src/assets/google-cloud.svg:1]: If you get this file on the Internet, could you please also attach a source and license?

File: cvat-ui/src/components/create-cloud-storage-page/cloud-storage-form.tsx [hunk 4]
```diff
@@ -39,13 +43,18 @@ interface CloudStorageForm {
...
     description?: string;
     region?: string;
+    prefix?: string;
+    project_id?: string;
     manifests: string[];
 }
 
 export default function CreateCloudStorageForm(props: Props): JSX.Element {
     const { cloudStorage } = props;
+    const cloudStorageId = (cloudStorage) ? cloudStorage.id : null;
```
Comment (bsekachev) [inline cvat-ui/src/components/create-cloud-storage-page/cloud-storage-form.tsx:57]:
```suggestion
    const cloudStorageId = cloudStorage ? cloudStorage.id : null;
```

File: cvat-ui/src/components/create-cloud-storage-page/cloud-storage-form.tsx [hunk 5]
```diff
@@ -170,15 +198,37 @@ export default function CreateCloudStorageForm(props: Props): JSX.Element {
                 duration: 15,
             });
         }
-    }, [cloudStorage]);
+    }, [cloudStorageId]);
```
Comment (bsekachev) [inline cvat-ui/src/components/create-cloud-storage-page/cloud-storage-form.tsx:201]: effect callback also depends on ``cloudStorage``

File: cvat-ui/src/components/create-cloud-storage-page/location.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    values: string[][];
+    href: string;
+}
+
+function prepareDefaultRegions(values: string[][]): Map<string, string> {
+    const temp = new Map<string, string>();
+    for (const [key, value] of values) {
+        temp.set(key, value);
+    }
+    return temp;
```
Comment (bsekachev) [inline cvat-ui/src/components/create-cloud-storage-page/location.tsx:32]: Object.fromEntries?

File: cvat-ui/src/components/create-cloud-storage-page/location.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                message: 'Incorrect region',
+                className: 'cvat-incorrect-add-region-notification',
+            });
+        } else if (regions.has(newRegionKey)) {
+            notification.warning({
+                message: 'This region already exists',
+                className: 'cvat-incorrect-add-region-notification',
+            });
+        } else {
+            const regionsCopy = regions;
```
Comment (bsekachev) [inline cvat-ui/src/components/create-cloud-storage-page/location.tsx:55]: I hardly think it is a copy, that is probably link to the same object. Check if ``regions`` contains ``newRegionKey`` after it was inserted to ``regionsCopy``.

Comment (bsekachev) [reply cvat-ui/src/components/create-cloud-storage-page/location.tsx:55]: If you need exactly copy, maybe better to use Object and create a shallow copy this way: ``const copy = { ...source }``

**Comment by bsekachev** [comment]
Looks like the patch works for me. But maybe one reccomendation. 
After I select a key file, the field is empty anyway
image

Probably better to show there file name.

**Comment by Marishka17** [comment]
@nmanovic, Could you please take a look at the server part of this PR?

File: cvat/apps/engine/models.py (hunk 1 above)
Comment (nmanovic) [reply cvat/apps/engine/models.py:600]: @Marishka17 , you can even have 255 (just in case). I really don't think that we have to be very strict here.

File: cvat/apps/engine/serializers.py [hunk 8]
```diff
@@ -809,6 +811,7 @@ class CloudStorageSerializer(serializers.ModelSerializer):
     key = serializers.CharField(max_length=20, allow_blank=True, required=False)
     secret_key = serializers.CharField(max_length=40, allow_blank=True, required=False)
     key_file_path = serializers.CharField(max_length=64, allow_blank=True, required=False)
```
Comment (nmanovic) [inline cvat/apps/engine/serializers.py:813]: @Marishka17 , I believe key_file_path has to be removed. What do you think? Probably you can do that in a separate PR.

**Review by nmanovic — APPROVED** [review]

File: cvat/apps/engine/serializers.py (hunk 8 above)
Comment (Marishka17) [reply cvat/apps/engine/serializers.py:802]: Hmm, I added the option to upload a key file for a case when CVAT instance is deployed remotely. But now I've looked at it from a different side. Ok, I'll prepare PR for this.
